File: src/637.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> averages;
        if (root == NULL) return averages;
        queue<TreeNode*> nodeQueues[2];
        nodeQueues[0].push(root);
        int level = 0, count = 0;
        double sum = 0.0;
        while (!nodeQueues[level].empty()) {
            vector<int> numsInLevel;
            level = 1 - level;
            sum = 0.0;
            count = 0;
            while (!nodeQueues[1 - level].empty()) {
                TreeNode* node = nodeQueues[1 - level].front();
                nodeQueues[1 - level].pop();
                sum += node->val;
                count++;
                if (node->left != NULL) nodeQueues[level].push(node->left);
                if (node->right != NULL) nodeQueues[level].push(node->right);
            }
            averages.push_back(sum / count);
        }
        return averages;
    }
};

int main() {
    Solution solution;
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    vector<double> ans = solution.averageOfLevels(root);
    for (auto it = ans.begin(); it != ans.end(); it++)
        cout << *it << " ";
    return 0;
}

File: pysrc/637.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def averageOfLevels(self, root: TreeNode) -> List[float]:
        if root is None:
            return []
        ans = []
        queue = [(root, 0)]
        currentDepth, sum, count = -1, 0, 1
        while len(queue) > 0:
            node, depth = queue.pop(0)
            if node is None:
                continue
            if depth != currentDepth:
                ans.append(sum / count)
                currentDepth, sum, count = depth, 0, 0

            sum, count = sum + node.val, count + 1
            queue.append((node.left, depth + 1))
            queue.append((node.right, depth + 1))

        if count != 0:
            ans.append(sum / count)

        return ans[1:]


if __name__ == '__main__':
    solution = Solution()
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    print(solution.averageOfLevels(root))
